### ISSUE TITLE
Implementing Stick Breaking Transform to transform real vector to simplex

### ISSUE THREAD

**Issue by luiarthur** [FluxML/Flux.jl#723]
Hi,

First I want to say thanks for your work on this package!

I am doing Bayesian modeling using variational inference, and am trying to implement a "Stick Breaking Transformation" to transform a vector of real numbers of dimension K to a simplex of dimension K+1.  The STAN team has documentation on this transformation [here][2]. And PyTorch implements it [here][1] (in the `_call` method of `StickBreakingTransform`). For convenience, I'll paste it here:

```python
def _call(self, x):
    offset = (x.shape[-1] + 1) - x.new([1]).expand(x.shape).cumsum(-1)
    z = torch.sigmoid(x - offset.log())
    z_cumprod = (1 - z).cumprod(-1)
    y = pad(z, (0, 1), value=1) * pad(z_cumprod, (1, 0), value=1)
    return y
```

I'm having some issues with implementing this using Flux. I think the issue here is with `vcat`, as I've seen other issues related to `vcat`. Not sure if in my case there's a workaround. In PyTorch, there's a `pad` function, which is used in their implementation of the stick breaking transformation.

Here's my current Julia implementation which uses Flux. It seems that I am not able to compute gradients of parameters correctly after using my implemented `stickBreakTransform`. Any thoughts on what may be the issue / workarounds?

```julia
using Flux, Flux.Tracker
using Flux.Tracker: TrackedArray, @grad, track

# Implement cumprod tracker
Base.cumprod(x::TrackedArray; dims=1) = track(cumprod, x, dims)
@grad cumprod(x::TrackedArray, dims) = 
    cumprod(x.data, dims=dims), g -> ( reverse(cumprod(reverse(g, dims=dims), dims=dims), dims=dims) , nothing)
 
"""
x: real vector of dim K - 1
return: simplex of dim K
"""
function stickBreakTransform(x::T) where T
  K = length(x) + 1
  k_vec = collect(1:K-1)
  z = sigmoid.(x .- log.(K .- k_vec))
  one_minus_z_cumprod = cumprod(1.0 .- z)

  # FIXME: I believe this line is the issue
  p = vcat(z, [1.0]) .* vcat([1.0], one_minus_z_cumprod)

  return p
end

### Test ###
y = param(ones(3))
p = stickBreakTransform(y)
z = sum(log.(p)) # -6.573399812223096 (tracked)
back!(z)

y.tracker.grad
# 3-element Array{Float64,1}:
# -35.216440811769075
# -17.937077242296343
# -3.018442006967813
```

Here is the same thing in PyTorch, which gives the correct result:
```python
import torch
from torch.distributions.transforms import StickBreakingTransform

sbt = StickBreakingTransform()
y = torch.ones(3, requires_grad=True)
p = sbt(y)
z = p.log().sum() # tensor(-6.5734, grad_fn=<SumBackward0>)
z.backward()

print(y.grad)
# tensor([-0.9015, -0.7284, -0.4621])
```

I'm using Julia v1.0.0 and Flux v0.8.1.

Thanks!

[1]: https://pytorch.org/docs/stable/_modules/torch/distributions/transforms.html
[2]: https://mc-stan.org/docs/2_19/reference-manual/simplex-transform-section.html

**Comment by luiarthur** [comment]
Sorry -- bad math on my part...

Nothing wrong with `vcat`. But my `cumprod` gradient is wrong (I copied and pasted the gradient for `cumsum`...). Not sure how to correctly implement gradient for `cumprod`. But, in the Stick Breaking Transform, where I only need cumulative products of positive numbers, I have the following workaround:

```julia
using Flux, Flux.Tracker
using Flux.Tracker: TrackedArray, @grad, track

# cumsum
Base.cumsum(x::TrackedArray; dims=1) = track(cumsum, x, dims)
@grad function cumsum(x::TrackedArray, dims)
  return cumsum(x.data, dims=dims), function(Δ)
    return (reverse(cumsum(reverse(Δ, dims=dims), dims=dims), dims=dims), nothing)
  end
end

# cumprod for positive numbers
cumprod_pos(x::T; dims=1) where T = exp.(cumsum(log.(x), dims=dims))
 
# TODO: general cumprod?
 
"""
x: real vector of dim K - 1
return: simplex of dim K
"""
function stickBreakTransform(x::T) where T
  K = length(x) + 1
  k_vec = collect(1:K-1)
  z = sigmoid.(x .- log.(K .- k_vec))
  one_minus_z_cumprod = cumprod_pos(1.0 .- z)

  # FIXME: I believe this line is the issue
  p = vcat(z, [1.0]) .* vcat([1.0], one_minus_z_cumprod)

  return p
end

### Test ###
y = param(ones(3))
p = stickBreakTransform(y)
z = sum(log.(p)) # -6.573399812223096 (tracked)
back!(z)

y.tracker.grad
# 3-element Array{Float
# -0.9014675456746869
# -0.7283506542974874
# -0.4621171572600099
```
which returns the expected results.

So, I'm closing this issue. Thanks!
